crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1101,6 +1118,67 @@ impl<'db> TypeInferenceBuilder<'db> {
             self.infer_definition(parameter);
         }
         self.infer_body(&function.body);
+
+        if let Some(declared_ty) = function
+            .returns
+            .as_deref()
+            .map(|ret| self.file_expression_type(ret))
+        {
+            fn is_suite_empty(suite: &[ast::Stmt]) -> bool {
+                let mut saw_ellipsis = false;
+                for statement in suite {
+                    if let Some(expr) = statement.as_expr_stmt() {
+                        if expr.value.is_ellipsis_literal_expr() {
+                            saw_ellipsis = true;
+                        } else if expr.value.is_string_literal_expr() {
+                            // docstring
+                        } else {
+                            return false;
+                        }
+                    }
+                }
+                saw_ellipsis
+            }
+            if is_suite_empty(&function.body) {
+                return;
+            }
+            let inferred_ty = if self.types.return_types_and_ranges.is_empty() {
+                KnownClass::NoneType.to_instance(self.db())
+            } else {
+                UnionType::from_elements(
+                    self.db(),
+                    self.types
+                        .return_types_and_ranges
+                        .iter()
+                        .map(|ty_range| ty_range.ty),
+                )
+            };
+            if !inferred_ty.is_assignable_to(self.db(), declared_ty) {
+                let mut reported = false;
+                for invalid in self
+                    .types
+                    .return_types_and_ranges
+                    .iter()
+                    .filter(|ty_range| !ty_range.ty.is_assignable_to(self.db(), declared_ty))
+                {
+                    report_invalid_return_type(
+                        &self.context,
+                        invalid.range,
+                        declared_ty,
+                        invalid.ty,
+                    );
+                    reported = true;
+                }
+                if !reported {
+                    report_invalid_return_type(
+                        &self.context,
+                        function.range(),
```

Comment:
Even for this case, reporting an error on the entire function definition node is not a good user experience, it leads to massive blocks of red squiggly lines in an editor. Pyright reports this error on the function return annotation node, I think that's a good choice.

Also, I think we should use a distinctive error message for this case. Something like "Function can implicitly return `None`, which is not assignable to return type {}"